Datasets
--------

So far we have compiled manifest files for the following publicly-available
studies:

* [ENCODE] ([papers][ENCODE papers])
* [GEUVADIS] ([papers][GEUVADIS papers])
* [Roadmap Epigenomics] ([papers][[Roadmap Epigenomics papers]])
* [BodyMap 2.0]
* [Pickrell et al] ([paper][Pickrell et al paper])
* [Cuffdiff 2] ([paper][Cuffdiff 2 paper])

[ENCODE]: http://www.genome.gov/10005107
[ENCODE papers]: http://www.genome.gov/10005107#al-2
[GEUVADIS]: http://www.geuvadis.org
[GEUVADIS papers]: http://www.geuvadis.org/web/geuvadis/publications
[Roadmap Epigenomics]: http://www.roadmapepigenomics.org
[Roadmap Epigenomics papers]: http://www.roadmapepigenomics.org/publications
[BodyMap 2.0]: http://www.ensembl.info/blog/2011/05/24/human-bodymap-2-0-data-from-illumina/
[Pickrell et al]: http://eqtl.uchicago.edu/Home.html
[Pickrell et al paper]: http://www.ncbi.nlm.nih.gov/pubmed/20220758
[Cuffdiff 2]: http://cufflinks.cbcb.umd.edu
[Cuffdiff 2 paper]: http://www.ncbi.nlm.nih.gov/pubmed/23222703

See `studies.tsv` for a more detailed description of the studies collected so
far and their respective sizes.  We're also hoping to include many other
studies that require us to apply for access, including:

* [TCGA]
* [Depression genes and networks]
* [METABRIC]
* [GTEx]

[TCGA]: https://tcga-data.nci.nih.gov/tcga/
[Depression genes and networks]: https://www.nimhgenetics.org/access_data_biomaterial.php
[METABRIC]: http://molonc.bccrc.ca/aparicio-lab/research/metabric/
[GTEx]: http://commonfund.nih.gov/GTEx/

As of 10/11/2013 all the manifest files in these subdirectories were created
by hand by me (Ben) using the best information I could find about the files
generated by these projects.  This is painstaking and error-prone and I hope,
but can't gaurantee that I haven't left any samples/files behind.

Data locations
--------------

There are multiple sequence read archives, the main one being NCBI's [SRA], and
then two others ([ENA] and [DDBJ]) that are essentially mirrors of NCBI's [SRA].

[SRA]: http://www.ncbi.nlm.nih.gov/sra
[ENA]: http://www.ebi.ac.uk/ena/
[DDBJ]: http://www.ddbj.nig.ac.jp

NCBI's [SRA] is problematic.  Their [file format] and URLs both change quite
often.  In addition, their [file format] (`.sra` / `.lite.sra`) is one that they
invented and can only be manipulated using their (shaky) tools.  Compressed
FASTQ is much easier to work with and, luckily, the other two archives store
reads as gzipped FASTQ, so I prefer to use URLs that point into those other
archives, not NCBI's.

[file format]: http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=std

Layout
------

Datasets are in intuitively-named subdirectories.  In a dataset directory,
you might find:

* [Myrna-style manifest file] called `manifest`
* Myrna-ready manifest file, usually called `manifest_mate1_randomized` or the
  like, which is like `manifest` but samples have been randomly divided into
  two groups, reflected in the new group label.  Also, all the mate-2 URLs
  have been removed, since Myrna can't handle paired-end reads.
* Python scripts used to generate the manifest file if I was lucky enough to
  find a way to do it automatically
* A script to run Myrna's preprocessor on the data, called `myrna_preprocess.sh`
* A script to run the rest of Myrna on the data, called `myrna.sh`

[Myrna-style manifest file]: http://bowtie-bio.sourceforge.net/myrna/manual.shtml#labeled-manifest-files
